### ISSUE TITLE
SMM: ListShardMapType.TryGetMappingForKey should support cached lookup

### ISSUE THREAD

**Issue by vyrotek** [Azure/elastic-db-tools#112]
It appears that GetMappings performs a lookup using the internal cache whereas looking for a specific mapping via TryGetMappingForKey does not.

I wasn't able to find this documented anywhere and it causes significant performance issues. This is inconvenient for systems that cannot use the OpenConnectionForKey method.

**Comment by jaredmoo** [comment]
Thanks for the feature request! We have been internally thinking about this problem and we have a few design ideas that I'd like to have an open discussion about. We'll post the design writeup when it's ready in the next few days.

To get the design discussion started could you describe your scenario, including what you do with the mappings after you get them and why you can't use OpenConnectionForKey?

**Comment by vyrotek** [comment]
We leverage both Entity Framework 6 (Using the designer/edmx) and Dapper for DB queries. As far as I know if we want to use sharding with EF this way we must build the connection string ourselves. We don't want to open a connection we won't use.

For example:

```
EntityConnectionStringBuilder entityBuilder = new EntityConnectionStringBuilder();
entityBuilder.Provider = "System.Data.SqlClient";            
entityBuilder.Metadata = @"res://*/Database.TheDatabase.csdl|...etc";
entityBuilder.ProviderConnectionString = connString; // From mapping.Shard info
var entityString = entityBuilder.ToString();
```

**Comment by jaredmoo** [comment]
That makes sense. 

Do your DB queries only ever operate on one mapping at a time? e.g. if keys 1, 2, and 3 are all mapped to shard1, when you connect to shard1 do you operate on keys 1, 2, and 3 on the same connection, or just one key at a time?

**Comment by jaredmoo** [comment]
Design proposal below. Naming of new types/values is provisional. There are some important design questions in bold. I will edit this post based on feedback.

vyrotek, I think that implementing the change in section 1 should satisfy your requirements.

# 1. Getting cached mappings

We will add a new `LookupOptions` with the below definition and new overloads to GetMappingForKey and TryGetMappingForKey methods. (**Q: do we need similar overloads for the various other GetMappings() methods?**) This will allow users to choose where to search for mapping data. 
* Passing `LookupOptions.None` will throw `ArgumentException`
* Passing `LookupOptions.LookupInCache` will look in cache only, and throw/return null if the mapping was not found (as per existing `GetMappingForKey` return/throw behavior).
* Passing `LookupOptions.LookupInStorage` will look in the backend storage only
* Passing `LookupOptions.LookupInCache | LookupOptions.LookupInStorage` will look in cache first, then look in storage on cache miss

```
[Flags]
public enum LookupOptions
{
    None = 0,
    LookupInCache = 1,
    LookupInStorage = 2,
}

public class ListShardMap<T>
{
...
    public PointMapping<TKey> GetMappingForKey() // existing method
    {
        return GetMappingForKey(LookupOptions.LookupInStorage);
    }
    
    public PointMapping<TKey> GetMappingForKey(LookupOptions lookupOptions) // new method
    {
        ...
    }
...
}

public class RangeShardMap<T>
{
...
    public RangeMapping<TKey> GetMappingForKey() // existing method
    {
        return GetMappingForKey(LookupOptions.LookupInStorage);
    }
    
    public RangeMapping<TKey> GetMappingForKey(LookupOptions lookupOptions) // new method
    {
        ...
    }
...
}
```

# 2. Skip repeatedly looking up unmapped shardlets
Some customers need the ability to have unmapped shardlets instead map to a 'default' shard. These apps will then occasionally (e.g. every 15 minutes) refresh the cache to pickup newly mapped keys. This can be implemented in the app as follows:

```
// Using ConcurrentDictionary as a Concurrent Hash Set, see http://stackoverflow.com/questions/18922985/concurrent-hashsett-in-net-framework
// Alternatively use a regular HashSet and use lock( )
private static ConcurrentDictionary<TKey, bool> unmappedKeyCache = ...

private IReadOnlyList<Shard> GetShards(IEnumerable<TKey> keys)
{
    // Gets the list of shards for these keys
    return keys.Select(k =>
    {
        if (unmappedKeyCache.Contains(k))
        {
            // We have cached that it's unmapped. Just return the default shard.
            return DefaultShard;
        }
        else
        {
            // Try to get the mapping
            RangeMapping<int> mapping;
            bool found = shardMap.TryGetMappingForKey(k, LookupOptions.LookupInCache | LookupOptions.LookupInStorage, out mapping);
            if (!found)
            {
                // Remember that this key is unmapped
                unmappedKeyCache.TryAdd(k, /* dummy value */ false);
                return DefaultShard;
            } 
            else
            {
                return mapping.ShardInfo;
            }
        }
    }).ToArray();
}

// Run the below every 15 minutes
private void ResetUnmappedCache()
{
    unmappedKeyCache.Clear();
}
```

# 3. Using cached mappings
Some customers need the ability to get a collection of cached mappings (as above) and then:
1. If the mappings point to one or more shard, connect to all of those shards (e.g. using MSQ) and use ExecuteReader()
2. If the mappings point to exactly one shard, connect to all of those shards (e.g. using MSQ or SqlConnection) and use ExecuteNonQuery().

Both of these must work correctly (i.e. no data corruption) when mappings change (i.e. are invalidated).

## 3.1 Create connection to multiple shards
To satisfy (1) above, we will add a new constructors `MultiShardConnection(IEnumerable<Shard>, ConnectionOptions)`. This will allow the user to write the below app code.

```
IEnumerable<TKey> keys = ...
IEnumerable<Shard> shards = GetShards(keys);    // uses cache
MultiShardConnection conn = new MultiShardConnection(shards, ConnectionOptions.Validate); // internally uses shard.OpenConnection(ConnectionOptions.Validate) to validate that Shard (and therefore all of its cached mappings) are up-to-date before connecting
```

The MultiShardConnection constructor should internally call `shards.Distinct()` to ensure that we only make one internal SqlConnection per shard.

What makes this correct is that when any mappings are added/modified/removed, the Shard's version id is also updated in the GSM and LSMs. This means that when we validate the Shards, we are also implicitly validating the Mappings that gave us those Shards.

**What do we do if our mapping cache contains mappings that have conflicting versions of the Shard? Then `new MultiShardConnection(shards, ConnectionOptions.Validate)` will throw for sure. What should app code look like to handle this?** 

## 3.2 Create connection to single shard
To satisfy (2) above, there are 2 options:

### 3.2 Option A - Expose shard.OpenConnection
Expose the `shard.OpenConnection(ConnectionOptions)` method. App code would look like below:

```
IEnumerable<TKey> keys = ...
IEnumerable<Shard> shards = GetShards(keys);    // uses cache
Shard shard = shards.Distinct().Single();       // throws exception if there is more than one Shard, or more than one version of the same Shard
SqlConnection conn = shard.OpenConnection(ConnectionOptions.Validate); // validate that Shard (and therefore all of its cached mappings) are up-to-date before connecting
```

**What do we do if our mapping cache contains mappings that have conflicting versions of the Shard? Then `new MultiShardConnection(shards, ConnectionOptions.Validate)` will throw for sure. What should app code look like to handle this?**

### 3.2 Option B - Allow MultiShardCommand.ExecuteNonQuery() if there is only one shard

Currently MultiShardCommand.ExecuteNonQuery() and MultiShardCommand.ExecuteScalar() are always blocked. It could be reasonable to allow these methods if we are only connecting to one shard. 

**What do we do if our mapping cache contains mappings that have conflicting versions of the Shard? Then `new MultiShardConnection(shards, ConnectionOptions.Validate)` will throw for sure. What should app code look like to handle this?**

**Comment by vyrotek** [comment]
Yes, I think these are great suggestions that would have helped make the behavior happening the hood more obvious.

Additionally, I also discovered that it **is** possible to pass an existing (and already open) SqlConnection to a new EF DbContext by putting it in a new EntityConnection and giving that to the DbContext constructor. I'm not sure how we missed this before but I feel like we were push towards the connection string way by following a few guides online. I apparently gave up too soon. I'm now able to use OpenConnectionForKey.

Hopefully this can help someone else work around TryGetMappingForKey not being cached.

**Comment by jaredmoo** [comment]
Cool, great to hear that your issue was resolved. Please don't hesitate to reach out on github issues or stackoverflow if you run into any other trouble. :)

**Comment by jaredmoo** [comment]
We discussed these options and decided to go ahead with the new API in section 1 above. We'll also continue providing the guidance in section 2 to customers who want to use that pattern (we will not add a new API specialized for it). The API design in section 3 is conceptually solves a different problem than what is raised in this issue so I'll open a new issue for it.

**Comment by adminnz** [comment]
Whats the status of this ? Because its been 5 months and GetMappingForKey & TryGetMappingForKey both still dot not utilize the cache for RangeShardMap<TKey>

The only current "workaround" which I'm not 100% sure is correct is to fetch the mappings and figure out which mapping the values should go into manually.
Example using the ElasticScaleStarterKit Shards.
```C#
RangeShardMap<int> shardMap = TryGetShardMap();
List<Int32> numbers = new List<int>();
int i = 1;
while (i < 200)
    numbers.Add(i++);
// Original: This makes a call to the database for each of the numbers
foreach (var group in numbers.GroupBy(n => shardMap.GetMappingForKey(n)))
{
Console.WriteLine("Range: {0} - {1}: {2:N0}", group.Key.Value.Low, group.Key.Value.High, group.Count());
}
//Manual: This makes a single call to get all the mappings, and then manually figures out the mapping for each number.
var mappings = shardMap.GetMappings();
foreach (var group in numbers.GroupBy(n => mappings.SingleOrDefault(r => r.Value.Low <= n && (r.Value.HighIsMax || r.Value.High > n))))
{
    Console.WriteLine("Range: {0} - {1}: {2:N0}", group.Key.Value.Low, group.Key.Value.High, group.Count());
}
```

Looking at the source code, the methods could be fixed by changing the line 
`bool result = this.rsm.TryLookup(key, false, out rangeMapping);` 
to 
`bool result = this.rsm.TryLookup(key, true, out rangeMapping);`

**Comment by jaredmoo** [comment]
I developed a prototype at https://github.com/jaredmoo/elastic-db-tools/tree/LookupFromCache but it is untested. I'll try to carve out some time to finish this work. In the meantime, please try it out and let me know if it solves your problem. Cheers

**Comment by jaredmoo** [comment]
Rebased https://github.com/jaredmoo/elastic-db-tools/tree/LookupFromCache on top of Azure master

**Comment by adminnz** [comment]
Well running my test against your prototype resulted in drastically reduced shard lookup times when using `LookupOptions.LookupInCache | LookupOptions.LookupInStore` options.

**Comment by jaredmoo** [comment]
That's great to hear. I really want to finish off this work and make an official build, but another product that I work on has critical work that I have to focus on for the next few weeks. Sorry again about the delay.

Just curious, what is your scenario that you need this for? I'm wondering if we can come up with another solution that will unblock you.

**Comment by adminnz** [comment]
Basically we have an Azure worker that downloads millions of status messages per day. 
The status messages normally have a timestamp around the current date +/- 2 weeks, and our databases are sharded based on that timestamp. (we drop old shards after 2-3 months)

We get a few thousand status messages each execution and we can't afford (Azure db's are not that fast) to do a single insert for each message so we have to batch them up run an insert using SqlBulkCopy.
But for it to work correctly we need to get the right shard each message should be going to, and then group them so that we only run one SqlBulkCopy per shard.

We could do the grouping logic ourselves but then it seams rather pointless to use the elastic-db-tools.

**Comment by jaredmoo** [comment]
Ok that makes sense - no way around it, cached lookup is indeed what you need to solve that problem :)

**Comment by adminnz** [comment]
So when is this going to get included into the main branch and released?

**Comment by jaredmoo** [comment]
Sorry, I can't provide any ETA at the moment. Is getting it merged into the main branch blocking you? If it's working well for you then hopefully you can continue to use a private build, I'm just hesitant to merge code that doesn't have any automated tests. If you are available to write a few tests I'd be willing to review and merge them. We need at least the following test cases: (1) ensure that `LookupOptions` are obeyed, i.e. only the specified locations are searched; and (2) ensure that if `LookupOptions.LookupInCache | LookupOptions.LookupInStore` is specified, the cache should be searched first.

**Comment by jaredmoo** [comment]
Released v2.1.0.
